Parallel point location

cjlh opened the pull request:
These changes update the cell location interface to accept a list of points to locate, rather than a single point.

A new PointCloud class has been created that can be used to store a list of points and to locate those points within the distributed mesh -- the results are given as an array of (rank, cell) pairs. A function may then be evaluated at these points on the mesh using the `evaluate` function (WIP).

wence- commented on firedrake/pointcloud.py:
```diff
@@ -0,0 +1,299 @@
+                i = local_candidate_indices[rank][j]
+                curr_rank = located_elements[i][0]
+                result = recv_results[rank][j][0]
+                if (result != -1 and (curr_rank == -1 or rank < curr_rank)):
```
I guess we could arbitrarily decide to choose the max rank, rather than the min rank, then we just write:
```
located_elements[i] = (max(rank, curr_rank), result)
```
With no need for the `if`.

wence- replied:
I think we never used the `point` here, so perhaps:
```python
for rank, result in recv_results.items():
   indices = local_candidate_indices[rank]
   found, = np.where(result != -1)
   for idx in found:
      i = indices[idx]
      loc_rank = located_elements[i, 0]
      if loc_rank == -1 or rank < loc_rank:
          located_elements[i, :] = (rank, result[idx])
```

cjlh replied:
I have went with your second suggestion here -- with the first, if two processes report locating the same point on different elements, would `located_elements` not then be updated with the new result regardless of which rank is higher or am I misunderstanding? Cheers.

cjlh commented on firedrake/pointcloud.py:
```diff
@@ -0,0 +1,299 @@
+class PointCloud(object):
+    """Store a group of points and evaluate the (cell, rank) pair corresponding their
+    location in the distributed mesh.
+    """
```
I'm happy with the description you've given here if others are too -- I agree it makes more sense.

cjlh commented on firedrake/pointcloud.py:
```diff
@@ -0,0 +1,299 @@
+        """Initialise the PointCloud.
+
+        :arg mesh: A mesh object.
+        :arg points: A list of points in the point cloud.
```
Sounds good to me.

cjlh commented on firedrake/pointcloud.py:
```diff
@@ -0,0 +1,299 @@
+        self.mesh = mesh
+        self.points = np.asarray(points)
+        _, dim = points.shape
+        assert dim == mesh.geometric_dimension()
```
That looks good to me.

cjlh commented on firedrake/pointcloud.py:
```diff
@@ -0,0 +1,299 @@
+        for i in range(0, self.mesh.comm.size):
+            j = arr_size * 2 * i
+            min_bounds.append(global_[j:j + arr_size])
+            max_bounds.append(global_[j + arr_size:j + arr_size * 2])
```
Thanks. I believe this should be:

```
min_bounds, max_bounds = global_.reshape(self.mesh.comm.size, 2, len(local) // 2).swapaxes(0, 1)
```

Without the calls to `flatten`, as `spatial_index.from_regions` expects a 2-dimensional array. The arrays must then also be made contiguous.

wence- replied:
Ah. So flatten makes contiguous arrays, you can then reshape afterwards.

cjlh commented on firedrake/pointcloud.py:
```diff
@@ -0,0 +1,299 @@
+        for rank, recv_buffers in recv_buffers_dict.items():
+            # Create all receive requests.
+            for i in range(0, len(recv_buffers)):
+                req = self.mesh.comm.Irecv(recv_buffers[i], source=rank, tag=i)
```
I think this becomes an issue when we use the same function for sending/receiving points (in this case `buffer` would be an array of n-dimensional arrays). As far as I can tell, mpi4py only allows size-1 arrays to be sent/received.

Noted on the naming -- thanks.

cjlh replied:
I think this becomes an issue when we use the same function for sending/receiving points (in this case `buffer` would be an array of n-dimensional arrays). As far as I can tell, mpi4py only allows size-1 arrays to be sent/received.

Noted on the naming -- thanks.

wence- replied:
I think we can send n-dimensional objects with mpi4py. But I could have misremembered.

cjlh replied:
I have now pushed the code to correct this and send the data as one message, cheers.

cjlh commented:
Thanks for the comments @wence-. I have implemented and pushed most of the changes you suggested -- happy to discuss further.

dham commented on firedrake/pointcloud.py:
```diff
@@ -0,0 +1,299 @@
+    """Store a group of points and evaluate the (cell, rank) pair corresponding their
+    location in the distributed mesh.
+    """
+    def __init__(self, mesh, points, *args, **kwargs):
```
I think that way around makes sense.

dham commented on firedrake/pointcloud.py:
```diff
@@ -0,0 +1,283 @@
+        """Initialise the PointCloud.
+
+        :arg mesh: A mesh object.
+        :arg points: An array of points.
```
What's a point? I think you mean an N x mesh.geometric_dimension() array of point locations.

dham commented on firedrake/pointcloud.py:
```diff
@@ -0,0 +1,283 @@
+        # Build spatial index of processes for point location.
+        self.processes_index = self._build_processes_spatial_index()
+
+    @cached_property
```
If this is a cached property, what is the mechanism for blowing the cache if the user moves the coordinate locations on the mesh?

dham replied:
Ideally to be safe, the `PointCloud` would register a callback on the mesh such that the cache gets blown every time `mesh.coordinate.dat.data` gets touched.

wence- replied:
I suppose that means we need to resurrect object versioning. I will add that to my lazy-evaluation purge.

dham commented on firedrake/pointcloud.py:
```diff
@@ -0,0 +1,283 @@
+        rank_cell_pairs = self._locate_mesh_elements()
+        return np.array(rank_cell_pairs, dtype=IntType)
+
+    def evaluate(self, function):
```
I'm not convinced that having an evaluate method here is wise. What happens if I want to use this in the Functional for an optimisation problem? What does the symbol look like for the derivative of this operator with respect to the point cloud? What does its adjoint look like?

I think a PointCloud is analogous to a FunctionSpace. We probably need PointFunction. Then you would have an `UFL.Delta(expr, cloud)` which would evaluate expr on the point cloud and gives us the right place to do all the derivatives and adjoints. 

I guess it would be OK to have an evaluate method as long as what it does is `return assemble(UFL.Delta(expr, self))`

wence- replied:
OK, let's punt on this for now (and remove the `evaluate` method). Instead a baby step in the right direction will be to make `Function.at` use the locations from here (rather than locating itself).

dham commented on firedrake/pointcloud.py:
```diff
@@ -0,0 +1,283 @@
+    def evaluate(self, function):
+        """Evaluate a function at the located points.
+
+        :arg function: The function to evaluate.
```
Either way I'm pretty sure you mean a UFL expression, not a function.

cjlh commented:
Thank you for the comments @dham, and for making the distinction between what I referred to as an array of "points" and N x mesh.geometric_dimension() array of point locations. I have corrected this and removed the evaluation method from the class.

wence- commented:
Reuben to look at in October.

lrtfm commented on firedrake/pointcloud.py:
```diff
@@ -0,0 +1,299 @@
+        :returns: A libspatialindex spatial index structure.
+        """
+        min_c = self.mesh.coordinates.dat.data_ro.min(axis=0)
+        max_c = self.mesh.coordinates.dat.data_ro.max(axis=0)
```
Should it be 
```
min_c = self.mesh.coordinates.dat.data_ro_with_halos.min(axis=0)
 max_c = self.mesh.coordinates.dat.data_ro_with_halos.max(axis=0)
```

ReubenHill commented:
Should this be closed? At a glance I think my `VertexOnlyMesh` work supersedes this
